### PR TITLE
Support brotli compressed boxes in JPEG XL

### PR DESCRIPTION
**Pull request by kmilos**
This introduces a dependency on the [brotli](https://github.com/google/brotli#introduction) library.

For some reason the Conan builds cannot find it, but the lib is found otherwise on macOS, MinGW, and Cygwin builds.

### CODE REVIEW

**Review by 1div0 — DISMISSED** [review]
OK

**Comment by postscript-dev** [comment]
@kmilos 
Is it worth updating [README.md](https://github.com/Exiv2/exiv2/blob/main/README.md) to reflect the new build option?

**Comment by kmilos** [comment]
@postscript-dev Good catch - will do at some point, and then you can comment on the changes.

The bigger stopper here is the Conan build problem, no ideas there...

**Review by 1div0 — DISMISSED** [review]
Gr8.

File: src/bmffimage.cpp [hunk 1]
```diff
@@ -452,6 +506,32 @@ uint64_t BmffImage::boxHandler(std::ostream& out /* = std::cout*/, Exiv2::PrintS
...
+      uint32_t realType = data.read_uint32(0, endian_);
+      if (bTrace) {
+        out << "type: " << toAscii(realType);
+      }
+#ifdef EXV_HAVE_BROTLI
+      DataBuf arr;
+      brotliUncompress(data.c_data(4), data.size() - 4, arr);
+      if (realType == TAG_exif) {
+        uint32_t offset = arr.read_uint32(0, endian_) + 4;
+        if (arr.size() > 8) {
```
Comment (kevinbackhouse) [inline src/bmffimage.cpp:520]:
```suggestion
        enforce(offset <= arr.size(), Exiv2::ErrorCode::kerCorruptedMetadata);
        if (arr.size() > 8) {
```

Comment (kmilos) [reply src/bmffimage.cpp:520]: Nice catch, thanks. I rolled in the min header length in there and got rid of the if clause while at it.

**Comment by kmilos** [comment]
@piponazo Do you have an idea about Conan problems? Is it related to case matching of files and variables? I saw Conan generates `FindBrotli.cmake` (no idea about its contents), while the working, hand-crafted one is `FindBROTLI.cmake`...

File: cmake/FindBrotli.cmake [hunk 2]
```diff
@@ -0,0 +1,45 @@
+#***************************************************************************
+#                                  _   _ ____  _
+#  Project                     ___| | | |  _ \| |
```
Comment (piponazo) [inline cmake/FindBrotli.cmake:3]: Was this file taken from the CURL project? Or is this a a copy/paste artifact?

File: cmake/findDependencies.cmake [hunk 3]
```diff
@@ -46,6 +46,10 @@ if( EXIV2_ENABLE_PNG )
     find_package( ZLIB REQUIRED )
 endif( )
 
+if( EXIV2_ENABLE_BMFF )
```
Comment (piponazo) [inline cmake/findDependencies.cmake:49]: [question] Do you always want to find the need library whenever BMFF is enabled? Are the two things tight together? Would it make sense to have a new CMake option for the support of the new library?

File: cmake/findDependencies.cmake [hunk 4]
```diff
@@ -30,6 +28,8 @@ else()
     list(APPEND CMAKE_PREFIX_PATH ${CMAKE_BINARY_DIR})
 endif()
 
+list(APPEND CMAKE_MODULE_PATH "${CMAKE_SOURCE_DIR}/cmake/")
```
Comment (piponazo) [inline cmake/findDependencies.cmake:31]: Why did you change the order of this line? I think that it was before the Conan section on purpose. Nonetheless, I would need to check this out with a computer. Right now I am reviewing the code from a tablet.

File: cmake/findDependencies.cmake [hunk 5]
```diff
@@ -46,6 +46,10 @@ if( EXIV2_ENABLE_PNG )
     find_package( ZLIB REQUIRED )
 endif( )
 
+if( EXIV2_ENABLE_BMFF )
+    find_package( Brotli )
```
Comment (piponazo) [inline cmake/findDependencies.cmake:50]: [question] If I understand this correctly, you are looking for the Brotli library optionally when BMFF is turned on, right?

**Review by piponazo — DISMISSED** [review]
I took a quick look to the changes from my tablet (I'm on vacations at the moment) and they look good.

**Comment by piponazo** [comment]
> @piponazo Do you have an idea about Conan problems? Is it related to case matching of files and variables? I saw Conan generates `FindBrotli.cmake` (no idea about its contents), while the working, hand-crafted one is `FindBROTLI.cmake`...

When I look at the checks from my tablet, I do not see any issues. Could you provide me a link to look at these issues?

What you describe is the normal behavior of Conan. The new Conan recipe that you are consuming will generate a `FindBrotli.CMake` in the build folder. In our CMake code we modify the CMAKE_PREFIX_PATH to include the `CMake` folder and the **build** in order to find such FindXXX.cmake files.

**Comment by kmilos** [comment]
Thanks @piponazo that was exactly it - I had to change the CMAKE_PREFIX_PATH list order so the Conan modules are searched for first, *and* make sure Conan's and the handcrafted module use the same letter case for variables.

All seems good now, so merging.

File: cmake/findDependencies.cmake (hunk 5 above)
Comment (kmilos) [reply cmake/findDependencies.cmake:50]: Yes.

File: cmake/findDependencies.cmake (hunk 4 above)
Comment (kmilos) [reply cmake/findDependencies.cmake:31]: Because the hand-made cmake/FindBrotli.cmake was being used, and that one doesn't know about Conan's install paths so it was not finding anything...

File: cmake/findDependencies.cmake (hunk 3 above)
Comment (kmilos) [reply cmake/findDependencies.cmake:49]: We could introduce an external switch as well... It is not a hard requirement for BMFF since it is (for now) specific only to JPEG XL, so I wanted to keep it simple.

File: cmake/FindBrotli.cmake (hunk 2 above)
Comment (kmilos) [reply cmake/FindBrotli.cmake:3]: Yes, taken from there, with some small modifications for letter case.

File: cmake/findDependencies.cmake (hunk 3 above)
Comment (kmilos) [reply cmake/findDependencies.cmake:49]: I think you have a point - I'll make it fully controllable so one has the option not to depend on it and still have (most of) the BMFF support.

File: cmake/findDependencies.cmake (hunk 5 above)
Comment (kmilos) [reply cmake/findDependencies.cmake]: Now changed to be controllable by user.

**Comment by kmilos** [comment]
Ok, now the CIFuzz is failing because brotli is on by default and now required: can someone update the docker image to include "libbrotli-dev" as well please, I have no idea how...? @piponazo @kevinbackhouse

File: .github/workflows/codeql-analysis.yml [hunk 6]
```diff
@@ -37,10 +37,15 @@ jobs:
     steps:
     - name: Checkout repository
       uses: actions/checkout@v3
+      
+    - name: Install dependencies
+      run: |
+        sudo apt-get update
+        sudo apt-get install -y libexpat1-dev zlib1g-dev libbrotli-dev
```
Comment (kevinbackhouse) [inline .github/workflows/codeql-analysis.yml:44]: I wonder how this worked before! expat and zlib are not new.

Comment (kmilos) [reply .github/workflows/codeql-analysis.yml:44]: They're installed by default!

Comment (kevinbackhouse) [reply .github/workflows/codeql-analysis.yml:44]: Another option would be to copy this:
 
https://github.com/Exiv2/exiv2/blob/b90426dda95c7b3f2824b6e137bf9f33820d36b4/.github/workflows/on_PR_linux_fuzz.yml#L26

**Comment by kevinbackhouse** [comment]
@kmilos: For CIFuzz, I think need to update this file:

https://github.com/google/oss-fuzz/blob/master/projects/exiv2/Dockerfile

Maybe we can change that to use `ci/install_dependencies.sh` so that it's less fragile in future.

File: .github/workflows/codeql-analysis.yml (hunk 6 above)
Comment (kmilos) [reply .github/workflows/codeql-analysis.yml:44]: I contemplated that, but seems to be overkill with all the gtest building etc.

**Comment by kmilos** [comment]
@kevinbackhouse Please go ahead, I'm running out of time/steam on this one...

File: src/bmffimage.cpp [hunk 7]
```diff
@@ -452,6 +506,31 @@ uint64_t BmffImage::boxHandler(std::ostream& out /* = std::cout*/, Exiv2::PrintS
...
+      uint32_t realType = data.read_uint32(0, endian_);
+      if (bTrace) {
+        out << "type: " << toAscii(realType);
+      }
+#ifdef EXV_HAVE_BROTLI
+      DataBuf arr;
+      brotliUncompress(data.c_data(4), data.size() - 4, arr);
+      if (realType == TAG_exif) {
+        uint32_t offset = arr.read_uint32(0, endian_) + 4;
+        enforce(offset + 4 < arr.size(), Exiv2::ErrorCode::kerCorruptedMetadata);
```
Comment (kevinbackhouse) [inline src/bmffimage.cpp:520]: The `offset + 4` could overflow here, allowing the `enforce` to be bypassed. I assume the purpose of this `enforce` is to prevent an integer overflow in `arr.size() - offset` below, so the `+4` looks unnecessary to me.
 
```suggestion
        enforce(offset < arr.size(), Exiv2::ErrorCode::kerCorruptedMetadata);
```

**Comment by kevinbackhouse** [comment]
@kmilos: https://github.com/google/oss-fuzz/pull/8835

File: src/bmffimage.cpp [hunk 8]
```diff
@@ -452,6 +506,31 @@ uint64_t BmffImage::boxHandler(std::ostream& out /* = std::cout*/, Exiv2::PrintS
...
+      enforce(data.size() >= 4, Exiv2::ErrorCode::kerCorruptedMetadata);
+      uint32_t realType = data.read_uint32(0, endian_);
+      if (bTrace) {
+        out << "type: " << toAscii(realType);
+      }
+#ifdef EXV_HAVE_BROTLI
+      DataBuf arr;
+      brotliUncompress(data.c_data(4), data.size() - 4, arr);
+      if (realType == TAG_exif) {
+        uint32_t offset = arr.read_uint32(0, endian_) + 4;
```
Comment (kevinbackhouse) [inline src/bmffimage.cpp:519]: This `+4` could overflow, but it wouldn't cause anything particularly bad to happen. So this suggestion is optional:

```suggestion
        uint32_t offset = Safe::add(arr.read_uint32(0, endian_), 4u);
```

File: src/bmffimage.cpp (hunk 7 above)
Comment (kmilos) [reply src/bmffimage.cpp:520]: The +4 is to account for the uint32 just read.

Comment (kevinbackhouse) [reply src/bmffimage.cpp:520]: In that case:

```suggestion
        enforce(Safe::add(offset, 4u) < arr.size(), Exiv2::ErrorCode::kerCorruptedMetadata);
```

**Comment by kmilos** [comment]
@kevinbackhouse I guess we have a chicken vs egg situation: the CIFuzz image still uses the install script from main, and not from this branch...

**Comment by kevinbackhouse** [comment]
@kmilos: No problem. We can merge without CIFuzz passing. Let me just check the latest changes before I approve.

**Review by kevinbackhouse — APPROVED** [review]
